docs/RFCS/20200414_scheduled_jobs.md

```diff
@@ -0,0 +1,424 @@
+* `job_args` is a serialized `ScheduledJobArguments` protocol message
+If next_run is `NULL`, then the job does not run (it's paused).
+* `executor_type` is a string identifying executor to use when executing this schedule
+* `executor_args` is a serialized `ExecutorArguments` protocol message
+* `schedule_changes` is a serialized `ScheduleChangeInfo` protocol message.
+
+Scheduled jobs system uses the `crontab` format to specify job schedule.  We believe that, even
+though this format is not the easiest to use/parse, the crontab format is, nonetheless,
+well-established and very flexible.  We will attempt to use existing crontab parsing
+packages if possible.
+
+The initial version of scheduled jobs will use UTC timezone for all the schedules.
+
+Scheduled jobs system uses `system.jobs` table as a table of record for the initiated scheduled
+job runs.  We will need to change the schema for the `system.jobs` table as follows:
+
+```sql
+ALTER TABLE system.jobs ADD COLUMN sched_id INT;
+CREATE INDEX sched_idx ON system.jobs(sched_id) STORING(status);
+```
+
+The `sched_id` indexed column references the schedule ID in the `system.scheduled_jobs`.
+We do not add `FOREIGN KEY` constraint (we want to keep historical records,
+even if we delete schedule)
+
+#### Protocol Messages vs Columns: side note
+It would be nice to put schedule in its own protocol message so that we may extend
+it in the future (e.g. add TZ).
+However, the use of protocol buffers directly in the database proved to be difficult
+from the UI perspective (the operator may want to view the schedule) as well as
+debug-ability.  If the native protobuf support improves, we may replace some
+of the `system.scheduled_jobs` fields with protocol messages.
+See [#47534](https://github.com/cockroachdb/cockroach/issues/47534)
+
+## Scheduled Job Execution
+Each node in the cluster runs a scheduled job execution daemon.
+At a high level, the cron daemon periodically performs the following steps:
+  1. Starts transaction
+  1. Polls `system.scheduled_jobs` to find eligible scheduled job
+  1. Executes the job
+      * Job execution means that we *queue* the job to be executed by the system jobs.
+      * Job execution is done by the job executor configured for this scheduled
+      * Job execution is subject to the cluster conditions (e.g. cluster settings);
+      * Future versions will also take into consideration cluster wide conditions 
+      (such as load), and may decide to reschedule the job instead of executing it.
+  1. Updates the `next_run` according to the cron experession, setting it to the next computed time,
+   or null if there is no next run.
+  1. Commits transaction 
+
+The daemon will apply a small amount of jitter to it's timing to reduce the
+probability of the conflicting transactions.
+
+The above steps run inside a single transaction to ensure that
+only one node starts running a job. Any errors encountered during
```

Comment:
I don't think two nodes would both run jobs for the same schedule -- since they'll create the job record in the same transaction as they eventually move `next_run`, they'll get a conflict and only one will win (the other will refresh, see next_run has changed and abort)

Reply:
Ah, right - I keep interpreting "executes the job" as actually running the job, rather than what it is actually doing which is just queuing the job for execution. Disregard my comment above.